src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,414 @@
+        return false;
+    }
+
+    if (!needUpdate(node_uuids, nodes_info))
+    {
+        LOG_TRACE(log, "No update required for cluster '{}'", cluster_info.name);
+        return true;
+    }
+
+    nodes_info = getNodes(zk, cluster_info.zk_root, node_uuids);
+    if (nodes_info.empty())
+    {
+        LOG_WARNING(log, "Can't get nodes info for '{}'", cluster_info.name);
+        return false;
+    }
+
+    int current_version;
+    getNodeNames(zk, cluster_info.zk_root, cluster_info.name, &current_version, true);
+
+    if (current_version != start_version)
+    {
+        LOG_TRACE(log, "Cluster '{}' configuration changed during update", cluster_info.name);
+        nodes_info.clear();
+        return false;
+    }
+
+    LOG_TRACE(log, "Updating system.clusters record for '{}' with {} nodes", cluster_info.name, cluster_info.nodes_info.size());
+
+    auto cluster = makeCluster(cluster_info);
+    context->setCluster(cluster_info.name, cluster);
+    return true;
+}
+
+bool ClusterDiscovery::updateCluster(const String & cluster_name)
+{
+    auto cluster_info = clusters_info.find(cluster_name);
+    if (cluster_info == clusters_info.end())
+    {
+        LOG_ERROR(log, "Unknown cluster '{}'", cluster_name);
+        return false;
+    }
+    return updateCluster(cluster_info->second);
+}
+
+void ClusterDiscovery::registerInZk(zkutil::ZooKeeperPtr & zk, ClusterInfo & info)
+{
+    LOG_DEBUG(log, "Registering current node {} in cluster {}", current_node_name, info.name);
+
+    String node_path = getShardsListPath(info.zk_root) / current_node_name;
+    zk->createAncestors(node_path);
+
+    zk->createOrUpdate(node_path, info.current_node.serialize(), zkutil::CreateMode::Ephemeral);
+    LOG_DEBUG(log, "Current node {} registered in cluster {}", current_node_name, info.name);
+}
+
+void ClusterDiscovery::start()
+{
+    if (clusters_info.empty())
+    {
+        LOG_DEBUG(log, "No defined clusters for discovery");
+        return;
+    }
+    LOG_TRACE(log, "Starting working thread");
+
+    auto zk = context->getZooKeeper();
+    for (auto & [_, info] : clusters_info)
+    {
+        registerInZk(zk, info);
+        if (!updateCluster(info))
+        {
+            LOG_WARNING(log, "Error on updating cluster '{}', will retry", info.name);
+            clusters_to_update->set(info.name);
+        }
+    }
+
+    main_thread = ThreadFromGlobalPool([this]
+    {
+        try
+        {
+            runMainThread();
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, "Caught exception in cluster discovery runMainThread");
+        }
+    });
+}
+
+void ClusterDiscovery::runMainThread()
+{
+    setThreadName("ClusterDiscover");
+    LOG_DEBUG(log, "Worker thread started");
+
+    using namespace std::chrono_literals;
+
+    while (!stop_flag)
+    {
+        /// if some cluster update was ended with error on previous iteration, we will retry after timeout
+        clusters_to_update->wait(5s);
```

Comment:
We check a node every 5 sec, so, probably, it is not vert critical to loose notification.
However, 5s is 5s.

Possible way to improve:
* add `size_t` version field to `clusters_to_update`, increment it inside `clusters_to_update->set`
* call `clusters_to_update->unset(cluster_name)` before `updateCluster(cluster_name)`
* in current loop check version of the cluster; condition in wait should compare version with the old one (from lst iteration)

Also, I'd better remove all atomics. Hash table does not change it's keys, and to shortly read/update value we can lock a mutex. There should not be a lot of contention.

Reply:
Also you can check stop_flag in the same condition (but lock the mutex when you set it)

Reply:
Discussed in PM that we can leave atomics, but add some global flag indicates if all clusters was updated or not not to wait before retry. This commit contains that changes [`434516c` (#31442)](https://github.com/ClickHouse/ClickHouse/pull/31442/commits/434516c72590b80c903d4da533116d50d104febb)